pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1139 @@
+                        },
+                        Err(e) => {
+                            // our caller is cancellation safe, but we might be racing with
+                            // another attempt to initialize. before we have cancellation
+                            // token support: these attempts should converge regardless of
+                            // their completion order.
+                            tracing::error!("layer file download failed, and additionally failed to communicate this to caller: {e:?}");
+                        }
+                    }
+                }
+
+                Ok(())
+            }
+            .in_current_span(),
+        );
+        match rx.await {
+            Ok(Ok(())) => {
+                if let Some(reason) = self
+                    .needs_download()
+                    .await
+                    .map_err(DownloadError::PostStatFailed)?
+                {
+                    // this is really a bug in needs_download or remote timeline client
+                    panic!("post-condition failed: needs_download returned {reason:?}");
+                }
+
+                self.consecutive_failures.store(0, Ordering::Relaxed);
+
+                Ok(())
+            }
+            Ok(Err(e)) => {
+                let consecutive_failures =
+                    self.consecutive_failures.fetch_add(1, Ordering::Relaxed);
+                tracing::error!(consecutive_failures, "layer file download failed: {e:#}");
+                let backoff = utils::backoff::exponential_backoff_duration_seconds(
+                    consecutive_failures.min(u32::MAX as usize) as u32,
+                    1.5,
+                    60.0,
+                );
+                let backoff = std::time::Duration::from_secs_f64(backoff);
+
+                // unless we get cancelled, we will hold off the init semaphore
+                tokio::time::sleep(backoff).await;
+
+                Err(DownloadError::DownloadFailed)
+            }
+            Err(_gone) => Err(DownloadError::DownloadCancelled),
+        }
+    }
+
+    async fn needs_download(&self) -> Result<Option<NeedsDownload>, std::io::Error> {
+        match tokio::fs::metadata(&self.path).await {
+            Ok(m) => Ok(self.is_file_present_and_good_size(&m)),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Some(NeedsDownload::NotFound)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn needs_download_blocking(&self) -> Result<Option<NeedsDownload>, std::io::Error> {
+        match self.path.metadata() {
+            Ok(m) => Ok(self.is_file_present_and_good_size(&m)),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Some(NeedsDownload::NotFound)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn is_file_present_and_good_size(&self, m: &std::fs::Metadata) -> Option<NeedsDownload> {
+        // in future, this should include sha2-256 the file, hopefully rarely, because info uses
```

Comment:
> Should the return type include the sha256? Should the function's execution also include sha2-256 validation?

I'd just validate it against the LayerFileMetadata, not spread the new hash anywhere.

> Option for the "error"

But it's not the error right? We look at the current state on disk (the std::fs::Metadata, later file contents) and deduce what needs to be done, if the contents needs to be redownloaded. This was however before the #5074 so a lot of the original ideas are a bit stale now.

Reply:
Oki, now understood what is `info`, it's the `Layer::info` method where we do the `is_remote` "check".

Reply:
> Instead, it returns a None if the file is present and has a good size, and it returns Some() if there is an "error" according to the name of the function. That's my concern here: the function name does not fit what is being returned. It can be resolved either by changing what is returned, or by changing the function name. I hope you understand :).

Well what would you then suggest?

Like `fn find_differences() -> Option<Differences>` sounds quite ok to me. I think the problem cannot be `fn needs_download() -> Option<NeedsDownload>`, because I can't find a better way to describe this.

So the problem is only the refactored common core around verifying a `std::fs::Metadata`:

```rust
fn is_file_present_and_good_size(&self, m: &std::fs::Metadata) -> Option<NeedsDownload> 
```

So this approach of `fn is_good() -> Option<Bad>` is confusing? You'd like to see `fn ensure_good() -> Result<(), Bad>`?

I am not really entirely convinced this was completly outside of what we can normally do, but still let's play the guessing game, since there was no suggestion on what to do.

```rust
fn is_file_present_and_good_size(&self, m: &std::fs::Metadata) -> Result<(), NeedsDownload>
```

Is this what you are after? How then do you change it to a function which returns an std::io::Error? Would you prefer something this:

```rust
enum NotGood {
  Fs(std::io::Error),
  NeedsDownload(NeedsDownload)
}
```

I quite dislike this approach compared to `Result<Option<NeedsDownload>, std::io::Error>` (what it is right now) because apart from `std::io::ErrorKind::NotFound` I cannot really handle error and need to propagate it up.

Reply:
55105ad1c